runtime/domain.c

```diff
@@ -417,50 +414,52 @@ static uintnat fresh_domain_unique_id(void) {
 /* must be run on the domain's thread */
 static void create_domain(uintnat initial_minor_heap_wsize) {
   dom_internal* d = 0;
+  caml_domain_state* domain_state;
+  struct interruptor* s;
+
   CAMLassert (domain_self == 0);
 
   /* take the all_domains_lock so that we can alter the STW participant
      set atomically */
   caml_plat_lock(&all_domains_lock);
 
   /* wait until any in-progress STW sections end */
-  while (atomic_load_acq(&stw_leader)) caml_plat_wait(&all_domains_cond);
+  while (atomic_load_acq(&stw_leader))
+    caml_plat_wait(&all_domains_cond);
 
   d = next_free_domain();
   if (d) {
-    struct interruptor* s = &d->interruptor;
+    s = &d->interruptor;
     CAMLassert(!s->running);
     CAMLassert(!s->interrupt_pending);
-    if (!s->interrupt_word) {
-      caml_domain_state* domain_state;
-      atomic_uintnat* young_limit;
-      /* never been started before, so set up minor heap */
-      if (!caml_mem_commit(
-              (void*)d->tls_area, (d->tls_area_end - d->tls_area))) {
-        /* give up now */
-        d = 0;
+
+    domain_self = d;
+
+    /* If the chosen domain slot has not been previously used, allocate a fresh
+     * domain state. Otherwise, reuse it. Reusing the slot ensures that the GC
+     * stats are not lost. */
+    if (!d->state) {
+      /* FIXME: Never freed. Not clear when to. */
+      domain_state = (caml_domain_state*)
+        caml_stat_calloc_noexc(1, sizeof(caml_domain_state));
+      if (domain_state == NULL)
```

Comment:
Minor note: this code uses `if (!d->state)` but also `if (domain_state == NULL)`: implicit-NULL-as-false and explicit-NULL-comparison. I think that the runtime house style is rather the latter, so `if (d->state == NULL)` would be more familiar to me.

(It's fine to leave the code as-is, but if you do make a change, maybe we could flip the test to check the `!= NULL` case first and have the short already-initialized branch out of the way?)